mindmeld/models/taggers/pytorch_crf.py

```diff
@@ -0,0 +1,408 @@
+
+        return self.inputs[index], mask
+
+
+def diag_concat_coo_tensors(tensors):
+    assert len(tensors) > 0
+
+    rows = []
+    cols = []
+    values = []
+    sparse_sizes = [0, 0]
+
+    nnz = 0
+    for tensor in tensors:
+        tensor = tensor.coalesce()
+        row, col = tensor.indices()[0], tensor.indices()[1]
+        if row is not None:
+            rows.append(row + sparse_sizes[0])
+
+        cols.append(col + sparse_sizes[1])
+
+        value = tensor.values()
+        if value is not None:
+            values.append(value)
+
+        sparse_sizes[0] += tensor.shape[0]
+        sparse_sizes[1] += tensor.shape[1]
+        nnz += tensor._nnz()
+
+    row = None
+    if len(rows) == len(tensors):
+        row = torch.cat(rows, dim=0)
+
+    col = torch.cat(cols, dim=0)
+
+    value = None
+    if len(values) == len(tensors):
+        value = torch.cat(values, dim=0)
+
+    return torch.sparse_coo_tensor(indices=torch.stack([row, col]), values=value, size=sparse_sizes).coalesce()
+
+
+def init_weights(m):
+    if isinstance(m, nn.Linear):
+        torch.nn.init.xavier_normal_(m.weight)
+        m.bias.data.fill_(0.01)
+
+
+def collate_tensors_and_masks(sequence):
+    if len(sequence[0]) == 3:
+        sparse_mats, masks, labels = zip(*sequence)
+        return diag_concat_coo_tensors(sparse_mats), torch.stack(masks), torch.stack(labels)
+    elif len(sequence[0]) == 2:
+        sparse_mats, masks = zip(*sequence)
+        return diag_concat_coo_tensors(sparse_mats), torch.stack(masks)
+
+
+class Encoder:
+    def __init__(self, feature_extractor="hash", num_feats=50000):
+
+        self.feat_extractor = DictVectorizer(dtype=np.float32) if feature_extractor == "dict" else FeatureHasher(
+            n_features=num_feats, dtype=np.float32)
+        self.label_encoder = LabelEncoder()
+        self.feat_extract_type = feature_extractor
+        self.num_classes = None
+        self.classes = None
+        self.num_feats = num_feats
+        self.ready = False
+
+    def get_tensor_data(self, feat_dicts, labels=None, fit=False):
+        if labels is None:
+            labels = []
+        if fit:
+            if self.feat_extract_type == "dict":
+                comb_dict_list = [x for seq in feat_dicts for x in seq]
+                self.feat_extractor.fit(comb_dict_list)
+                self.num_feats = len(self.feat_extractor.get_feature_names())
+            if labels:
+                self.label_encoder.fit([x for l in labels for x in l])
+                self.pad_index = len(self.label_encoder.classes_) - 1
+                self.classes = self.label_encoder.classes_
+                self.num_classes = len(self.label_encoder.classes_)
+
+            self.ready = True
+        feats = []
+        encoded_labels = []
+        seq_lens = [len(x) for x in feat_dicts]
+        max_seq_len = max(seq_lens)
+
+        for i, (x, y) in enumerate(zip_longest(feat_dicts, labels)):
```

Comment:
it's unclear from the context what the shape of `x` is

Reply:
Refactored code and added comment to make it more clear